src/libmongoc/tests/test-mongoc-error.c

```diff
@@ -133,6 +134,52 @@ test_has_label (void)
    BSON_ASSERT (!mongoc_error_has_label (tmp_bson ("{}"), "foo"));
 }
 
+static void
+test_state_change_helper (uint32_t domain, bool expect_error)
+{
+   bson_error_t error;
+   uint32_t not_master_codes[] = {10107, 13435};
+   uint32_t node_is_recovering_codes[] = {11600, 11602, 13436, 189, 91};
+   uint32_t shutdown_codes[] = {11600, 91};
+   int i;
+
+   MONGOC_DEBUG ("Checking domain = %d", domain);
+
+   memset (&error, 0, sizeof (bson_error_t));
+   error.domain = domain;
+
+   for (i = 0; i < sizeof (not_master_codes) / sizeof (uint32_t); i++) {
+      error.code = not_master_codes[i];
+      BSON_ASSERT (expect_error == _mongoc_error_is_not_master (&error));
+      BSON_ASSERT (expect_error == _mongoc_error_is_state_change (&error));
+   }
+   for (i = 0; i < sizeof (node_is_recovering_codes) / sizeof (uint32_t); i++) {
+      error.code = node_is_recovering_codes[i];
+      BSON_ASSERT (expect_error == _mongoc_error_is_state_change (&error));
+   }
+   for (i = 0; i < sizeof (shutdown_codes) / sizeof (uint32_t); i++) {
+      error.code = shutdown_codes[i];
+      BSON_ASSERT (expect_error == _mongoc_error_is_shutdown (&error));
+   }
+
+   error.code = 123;
+   bson_strncpy (error.message, "... not master ...", sizeof (error.message));
+   BSON_ASSERT (expect_error == _mongoc_error_is_not_master (&error));
```

Comment:
I think that if `error.message = " ... not master or secondary ..."` and `error.code` is left as is then  `_mongoc_error_is_not_master (&error)` should return false according to the spec since `_error_is_recovering` will return true:
```
def isNotMaster(message, code):
    if code and code in notMasterCodes:
        return true
    # if no code or an unrecognized code, use the error message.
    if isRecovering(message, None):
        return false
    return ("not master" in message)
```
I don't think that's accounted for here but I could be missing something.

Reply:
Agreed, added the test.